File: packages/js/src/presets.ts
import { HelixField } from "./field";
import type {
  HelixNoiseOptions,
  Bake2DResult,
  Bake3DResult,
  BoundaryOptions,
  BoundedField,
  Field,
  FlowField,
  Out6,
  ScaleFn,
  Sdf,
  Vec3,
} from "./types";
import { BoundedFieldImpl } from "./boundary";
import { lambda2FromGrad, qFromGrad, stretchFromGrad } from "./diagnostics";

/**
 * Ready-made shapes for the per-wavenumber dials (`spectrum`, `coherence`, `helicity`) plus two
 * field factories. Everything here is a pure function of `|k|` or a closed form — no RNG, so
 * mixing presets in never disturbs a field's mode layout.
 */

/**
 * Spectrum preset: a Gaussian shell bump `a(k) = exp(−(k−kPeak)²/(2·width²))` — an
 * energy-containing band instead of a power law. Pair it with `kmin ≈ kPeak − 3·width` and
 * `kmax ≈ kPeak + 3·width` so the discarded tails stay near 1%.
 *
 * ```js
 * create({ spectrum: shellPeak(3), kmin: 1, kmax: 6 });
 * ```
 */
export function shellPeak(kPeak: number, width = 1): ScaleFn {
  const w = Math.abs(width) || 1;
  return (k: number): number => Math.exp((-(k - kPeak) * (k - kPeak)) / (2 * w * w));
}

/**
 * Coherence preset: `λ(k) = clamp(1 − k/kc, 0, 1)` — organized structure at large scales,
 * fading to uncorrelated noise above the cutoff `kc`. The "rollers carrying fine grain" look.
 */
export function rolloff(kc: number): ScaleFn {
  return (k: number): number => Math.min(1, Math.max(0, 1 - k / kc));
}

/**
 * Helicity preset: `p(k) = k <= kSplit ? pLarge : pSmall` — a handed large-scale condensate
 * over a (near-)mirror-symmetric fine-scale background, the polarization profile measured on
 * forced turbulence.
 */
export function condensate(kSplit: number, pLarge: number, pSmall = 0): ScaleFn {
  return (k: number): number => (k <= kSplit ? pLarge : pSmall);
}

/**
 * The classical ABC (Arnold–Beltrami–Childress) cell field
 * `u = (A sin z + C cos y, B sin x + A cos z, C sin y + B cos x)`,
 * built as exactly three modes of the same engine — so `glsl()`, the bakes, `withBoundary`
 * and every sampler work on it unchanged. Consumes no RNG, is exactly 2π-tileable, and is a
 * pure Beltrami field (`∇×u = u`, potential `A = u`).
 *
 * By default this is the *literal* field (its amplitudes mean what they say). Pass
 * `amplitude` to opt into the usual RMS normalization, or `decay: ν` for the exact viscous
 * solution `u(t) = e^(−νt)·u(0)`.
 */
export function abc(A = 1, B = 1, C = 1, opts?: { amplitude?: number; decay?: number }): Field {
  // frame() maps these three directions to the (e1, e2) pairs that make the mode formula
  // a(cos φ·e1 − s·sin φ·e2) reproduce each ABC term; the phases below place sin/cos.
  const HALF_PI = Math.PI / 2;
  // Closed-form RMS: the 5³ grid mean of |u|² is exactly (A²+B²+C²) in exact arithmetic, so
  // never route this through rms() — float summation order would move it by ~1 ulp.
  const rms = Math.sqrt(A * A + B * B + C * C);
  const amp = opts?.amplitude;
  return new HelixField(
    { modes: 3, tileable: true, churn: 0, decay: opts?.decay ?? 0 },
    {
      k: [
        [0, 0, 1],
        [1, 0, 0],
        [0, 1, 0],
      ],
      s: [1, 1, 1],
      a: [A, B, C],
      ph: [-HALF_PI, -HALF_PI, Math.PI],
      scale: amp === undefined ? 1 : amp / (rms || 1),
    }
  );
}

/** Detail-amplitude constant for {@link twoScale}: `amplitude = C_TWO_SCALE / kDetail` holds the
 *  detail layer's vorticity budget fixed as you move its wavenumber. */
export const C_TWO_SCALE = 1.6;

/**
 * The sum of two divergence-free fields — still divergence-free, and its potential is the sum
 * of theirs, so `withBoundary` and the potential bakes keep working.
 *
 * The recipe it exists for: a coherent large-scale backbone plus incoherent broadband detail.
 * Sizing the detail as `amplitude: C_TWO_SCALE / kDetail` keeps its vorticity budget constant
 * while you slide the detail scale.
 *
 * ```js
 * const kD = 8;
 * const detail = create({ spectrum: shellPeak(kD), kmin: kD - 3, kmax: kD + 3,
 *                         amplitude: C_TWO_SCALE / kD, seed: 2 });
 * const storm = twoScale(abc(3, 3, 3), detail);
 * ```
 */
export function twoScale(base: FlowField, detail: FlowField, opts?: { detailGain?: number }): FlowField {
  return new TwoScaleField(base, detail, opts?.detailGain ?? 1);
}

/** Componentwise sum of two flow fields. @internal */
class TwoScaleField implements FlowField {
  /**
   * Per-instance, every one of them. `twoScale` is a sum node exactly as `compose` is, so a part
   * that is itself a `TwoScaleField` would otherwise be handed the same arrays its caller is
   * mid-sum on. Nesting in the `detail` slot is the visible case — the inner call overwrites `_a6`
   * with its own base, so the outer field's base is dropped and the inner one counted twice — while
   * nesting in `base` survives only because that slot happens to be scaled in place. A sum whose
   * correctness depends on which side you nested is not one to leave standing.
   */
  private readonly _a6: number[] = [0, 0, 0, 0, 0, 0];
  private readonly _b6: number[] = [0, 0, 0, 0, 0, 0];
  private readonly _t6: number[] = [0, 0, 0, 0, 0, 0];
  private readonly _a9: number[] = [0, 0, 0, 0, 0, 0, 0, 0, 0];
  private readonly _b9: number[] = [0, 0, 0, 0, 0, 0, 0, 0, 0];
  private readonly _g9: number[] = [0, 0, 0, 0, 0, 0, 0, 0, 0];

  constructor(
    readonly base: FlowField,
    readonly detail: FlowField,
    readonly detailGain: number
  ) {}

  private _sum<T extends Out6>(
    pick: "uw" | "ua",
    x: number,
    y: number,
    z: number,
    out6: T,
    t: number
  ): T {
    const g = this.detailGain;
    if (pick === "uw") {
      this.base.sampleUW(x, y, z, this._a6, t);
      this.detail.sampleUW(x, y, z, this._b6, t);
    } else {
      this.base.sampleUA(x, y, z, this._a6, t);
      this.detail.sampleUA(x, y, z, this._b6, t);
    }
    for (let i = 0; i < 6; i++) out6[i] = this._a6[i] + g * this._b6[i];
    return out6;
  }

  sampleUW<T extends Out6>(x: number, y: number, z: number, out6: T, t = 0): T {
    return this._sum("uw", x, y, z, out6, t);
  }

  sampleUA<T extends Out6>(x: number, y: number, z: number, out6: T, t = 0): T {
    return this._sum("ua", x, y, z, out6, t);
  }

  /** Both scales are analytic, so their weighted sum is too. */
  sampleGrad<T extends Out6>(x: number, y: number, z: number, out9: T, t = 0): T {
    if (out9.length < 9) throw new Error("helix-noise: sampleGrad needs 9 floats");
    this.base.sampleGrad(x, y, z, this._a9, t);
    this.detail.sampleGrad(x, y, z, this._b9, t);
    const g = this.detailGain;
    for (let i = 0; i < 9; i++) out9[i] = this._a9[i] + g * this._b9[i];
    return out9;
  }

  qCriterion(x: number, y: number, z: number, t = 0): number {
    return qFromGrad(this.sampleGrad(x, y, z, this._g9, t));
  }

  lambda2(x: number, y: number, z: number, t = 0): number {
    return lambda2FromGrad(this.sampleGrad(x, y, z, this._g9, t));
  }

  stretching(x: number, y: number, z: number, t = 0): number {
    this.sampleUW(x, y, z, this._t6, t);
    return stretchFromGrad(this.sampleGrad(x, y, z, this._g9, t), this._t6[3], this._t6[4], this._t6[5]);
  }

  sample(x: number, y: number, z: number, t = 0): Vec3 {
    this.sampleUW(x, y, z, this._t6, t);
    return [this._t6[0], this._t6[1], this._t6[2]];
  }

  vorticity(x: number, y: number, z: number, t = 0): Vec3 {
    this.sampleUW(x, y, z, this._t6, t);
    return [this._t6[3], this._t6[4], this._t6[5]];
  }

  helicityDensity(x: number, y: number, z: number, t = 0): number {
    this.sampleUW(x, y, z, this._t6, t);
    return this._t6[0] * this._t6[3] + this._t6[1] * this._t6[4] + this._t6[2] * this._t6[5];
  }

  potential(x: number, y: number, z: number, t = 0): Vec3 {
    this.sampleUA(x, y, z, this._t6, t);
    return [this._t6[3], this._t6[4], this._t6[5]];
  }

  withBoundary(sdf: Sdf, opts?: BoundaryOptions): BoundedField {
    return new BoundedFieldImpl(this, sdf, opts);
  }

  bake3D(n: number, t = 0): Bake3DResult {
    const data = new Float32Array(n * n * n * 4);
    let p = 0;
    for (let z = 0; z < n; z++) for (let y = 0; y < n; y++) for (let x = 0; x < n; x++) {
      this.sampleUW((x / n) * TAU_, (y / n) * TAU_, (z / n) * TAU_, this._t6, t);
      data[p] = this._t6[0]; data[p + 1] = this._t6[1]; data[p + 2] = this._t6[2];
      data[p + 3] = this._t6[0] * this._t6[3] + this._t6[1] * this._t6[4] + this._t6[2] * this._t6[5];
      p += 4;
    }
    return { data, size: n, channels: 4 };
  }

  bakePotential3D(n: number, t = 0): Bake3DResult {
    const data = new Float32Array(n * n * n * 4);
    let p = 0;
    for (let z = 0; z < n; z++) for (let y = 0; y < n; y++) for (let x = 0; x < n; x++) {
      const px = (x / n) * TAU_, py = (y / n) * TAU_, pz = (z / n) * TAU_;
      this.sampleUA(px, py, pz, this._t6, t);
      data[p] = this._t6[3]; data[p + 1] = this._t6[4]; data[p + 2] = this._t6[5];
      this.sampleUW(px, py, pz, this._t6, t);
      data[p + 3] = this._t6[0] * this._t6[3] + this._t6[1] * this._t6[4] + this._t6[2] * this._t6[5];
      p += 4;
    }
    return { data, size: n, channels: 4 };
  }

  bake2D(nx: number, ny: number, z = 0, t = 0): Bake2DResult {
    const data = new Float32Array(nx * ny * 4);
    let p = 0;
    for (let j = 0; j < ny; j++) for (let i = 0; i < nx; i++) {
      this.sampleUW((i / nx) * TAU_, (j / ny) * TAU_, z, this._t6, t);
      data[p] = this._t6[0]; data[p + 1] = this._t6[1]; data[p + 2] = this._t6[2];
      data[p + 3] = this._t6[0] * this._t6[3] + this._t6[1] * this._t6[4] + this._t6[2] * this._t6[5];
      p += 4;
    }
    return { data, width: nx, height: ny, channels: 4 };
  }
}

const TAU_ = 2 * Math.PI;

/** Options for {@link exactNS}. */
export interface ExactNSOptions {
  /** Shell wavenumber k₀ (used for both `kmin` and `kmax`). Default 2. */
  k0?: number;
  /** Kinematic viscosity ν (becomes `decay`). Default 0.02; `0` gives the steady Euler member. */
  nu?: number;
  /** Chirality s = ±1 (becomes `helicity`, so every mode takes that sign). Default 1. */
  sign?: 1 | -1;
  /** Passed through untouched — none of these break exactness. */
  modes?: number;
  seed?: number;
  amplitude?: number;
  coherence?: number;
  centers?: number;
  layout?: "fibonacci" | "random";
}

/**
 * Options for a field that is an **exact solution of the Navier–Stokes equations**, not merely a
 * plausible-looking one: a single-wavenumber Beltrami field decaying at the exact Stokes rate.
 *
 * All modes share one `|k| = k₀` and one handedness, which makes the field Beltrami
 * (`∇×u = ±k₀·u`); the nonlinear term is then a pure gradient and is absorbed by the pressure, so
 * the flow just decays as `e^(−νk₀²t)` — which is precisely what `decay` does to every amplitude.
 *
 * ```js
 * const field = create(exactNS({ k0: 2, nu: 0.05, seed: 7 }));
 * ```
 *
 * Keep `tileable` off and `ellipticity` at 1 — integer rounding would break the single shell, and
 * elliptic modes are not Beltrami.
 */
export function exactNS(opts: ExactNSOptions = {}): HelixNoiseOptions {
  const { k0 = 2, nu = 0.02, sign = 1, ...rest } = opts;
  return {
    kmin: k0,
    kmax: k0, // one shell: |k| is the same for every mode
    helicity: sign, // p = ±1 makes every drawn sign that sign
    churn: 0, // om = 0; the only time dependence is the viscous decay
    decay: nu,
    tileable: false,
    ellipticity: 1, // Beltrami — the whole point
    ...rest,
  };
}

/**
 * Energy-weighted polarization of the two measured turbulence states these presets aim at:
 * `d` = linear polarization degree, `absP` = per-shell helical fraction, `signedP` = its signed
 * mean. Quoted so tests and docs can check what a field actually reproduces.
 */
export const NS_TARGETS = {
  dev: { d: 1.155, absP: 0.462, signedP: 0.057 },
  forced: { d: 1.062, absP: 0.516, signedP: 0.229, condensateK1: 0.55 },
} as const;

/** Invert the per-mode relative helicity 2ε/(1+ε²) = p for ε — the ellipticity matching a target. */
function epsForP(p: number): number {
  return (1 - Math.sqrt(1 - p * p)) / p;
}

/**
 * Option bundle whose **polarization** matches measured developed turbulence: per-mode helical
 * fraction |p| ≈ 0.46 with a weakly positive signed mean, i.e. strongly polarized waves whose
 * handedness nearly cancels. Both numbers are inverted exactly — `ellipticity` from
 * 2ε/(1+ε²) = |p|, and the `helicity` slider from signedP/|p|.
 *
 * Scope, honestly: the polarization is calibrated, the **spectrum is not** — this keeps the
 * default power law. Matching the measured shell spectrum needs a table this package does not
 * ship; pass your own `spectrum` if you have one.
 */
export function nsDeveloped(overrides?: Partial<HelixNoiseOptions>): HelixNoiseOptions {
  const { absP, signedP } = NS_TARGETS.dev;
  return {
    modes: 96,
    layout: "random", // a genuine ensemble, matching how the target was measured
    kmin: 1,
    kmax: 15, // the band holding ~99% of the measured energy
    ellipticity: epsForP(absP),
    helicity: signedP / absP,
    ...overrides,
  };
}

/**
 * Same as {@link nsDeveloped} for the *forced* state: more strongly polarized waves (|p| ≈ 0.52)
 * and a markedly net-handed field (signed mean +0.23), which reads as a visible overall swirl
 * direction. The measured state also carries a helical condensate at the largest scale; expressing
 * that needs a per-wavenumber `helicity` — see the `condensate` preset if you want to add one.
 *
 * Same scope caveat as {@link nsDeveloped}: polarization calibrated, spectrum not.
 */
export function nsForced(overrides?: Partial<HelixNoiseOptions>): HelixNoiseOptions {
  const { absP, signedP } = NS_TARGETS.forced;
  return {
    modes: 96,
    layout: "random",
    kmin: 1,
    kmax: 15,
    ellipticity: epsForP(absP),
    helicity: signedP / absP,
    ...overrides,
  };
}

File: packages/rust/src/presets.rs
//! Ready-made shapes for the per-wavenumber dials, plus two field factories.
//!
//! Everything here is a pure function of `|k|` or a closed form — no RNG, so mixing presets in
//! never disturbs a field's mode layout.

use crate::boundary::VectorPotential;
use crate::constants::{HelixOptions, ScaleFn};
use crate::field::HelixField;

/// Spectrum preset: a Gaussian shell bump `a(k) = exp(-(k-k_peak)^2 / (2*width^2))` — an
/// energy-containing band instead of a power law. Pair it with `kmin ≈ k_peak - 3*width` and
/// `kmax ≈ k_peak + 3*width` so the discarded tails stay near 1%.
pub fn shell_peak(k_peak: f64, width: f64) -> ScaleFn {
    let w = if width.abs() == 0.0 { 1.0 } else { width.abs() };
    Box::new(move |k: f64| (-(k - k_peak) * (k - k_peak) / (2.0 * w * w)).exp())
}

/// Coherence preset: `lambda(k) = clamp(1 - k/kc, 0, 1)` — organized structure at large scales,
/// fading to uncorrelated noise above the cutoff `kc`.
pub fn rolloff(kc: f64) -> ScaleFn {
    Box::new(move |k: f64| (1.0 - k / kc).clamp(0.0, 1.0))
}

/// Helicity preset: `p(k) = if k <= k_split { p_large } else { p_small }` — a handed large-scale
/// condensate over a (near-)mirror-symmetric fine-scale background.
pub fn condensate(k_split: f64, p_large: f64, p_small: f64) -> ScaleFn {
    Box::new(move |k: f64| if k <= k_split { p_large } else { p_small })
}

/// Options for [`abc`].
#[derive(Default)]
pub struct AbcOptions {
    /// When set, RMS-normalize to this amplitude (via the closed form `sqrt(A²+B²+C²)`).
    /// When `None` (the default) the literal field is returned.
    pub amplitude: Option<f64>,
    /// Viscosity `nu`: gives the exact viscous solution `u(t) = e^(-nu t) u(0)`.
    pub decay: f64,
}

/// The classical ABC (Arnold–Beltrami–Childress) cell field
/// `u = (A sin z + C cos y, B sin x + A cos z, C sin y + B cos x)`, built as exactly three
/// modes of the same engine — so `glsl()`, the bakes, `with_boundary` and every sampler work
/// on it unchanged. Consumes no RNG, is exactly 2π-tileable, and is a pure Beltrami field
/// (`curl u = u`, potential `A = u`).
pub fn abc(a: f64, b: f64, c: f64, opts: AbcOptions) -> HelixField {
    use std::f64::consts::PI;
    // Closed-form RMS: the 5³ grid mean of |u|² is exactly A²+B²+C² in exact arithmetic, so
    // never route this through rms() — float summation order would move it by ~1 ulp.
    let rms = (a * a + b * b + c * c).sqrt();
    let scale = match opts.amplitude {
        None => 1.0,
        Some(amp) => amp / if rms == 0.0 { 1.0 } else { rms },
    };
    HelixField::from_modes(
        &[[0.0, 0.0, 1.0], [1.0, 0.0, 0.0], [0.0, 1.0, 0.0]],
        &[1.0, 1.0, 1.0],
        &[a, b, c],
        &[-PI / 2.0, -PI / 2.0, PI],
        scale,
        opts.decay,
    )
}

/// Detail-amplitude constant for [`TwoScale`]: `amplitude = C_TWO_SCALE / k_detail` holds the
/// detail layer's vorticity budget fixed as you move its wavenumber.
pub const C_TWO_SCALE: f64 = 1.6;

/// A coherent large-scale backbone plus incoherent broadband detail: the componentwise sum of
/// two divergence-free fields, which is divergence-free, and whose potential is the sum of
/// theirs (so boundaries keep working).
pub struct TwoScale<'a> {
    /// The large-scale backbone.
    pub base: &'a HelixField,
    /// The fine-scale layer.
    pub detail: &'a HelixField,
    /// Multiplier on the detail layer's output.
    pub detail_gain: f64,
}

impl<'a> TwoScale<'a> {
    /// Compose two fields; `detail_gain` scales the detail layer (1.0 = as built).
    pub fn new(base: &'a HelixField, detail: &'a HelixField, detail_gain: f64) -> Self {
        TwoScale { base, detail, detail_gain }
    }

    /// Velocity `u` and vorticity `w` of the composite.
    pub fn sample_uw(&self, x: f64, y: f64, z: f64, t: f64) -> ([f64; 3], [f64; 3]) {
        let (bu, bw) = self.base.sample_uw(x, y, z, t);
        let (du, dw) = self.detail.sample_uw(x, y, z, t);
        let g = self.detail_gain;
        let mut u = [0.0; 3];
        let mut w = [0.0; 3];
        for i in 0..3 {
            u[i] = bu[i] + g * du[i];
            w[i] = bw[i] + g * dw[i];
        }
        (u, w)
    }

    /// Velocity `u` and analytic vector potential `A` of the composite.
    pub fn sample_ua(&self, x: f64, y: f64, z: f64, t: f64) -> ([f64; 3], [f64; 3]) {
        let (bu, ba) = self.base.sample_ua(x, y, z, t);
        let (du, da) = self.detail.sample_ua(x, y, z, t);
        let g = self.detail_gain;
        let mut u = [0.0; 3];
        let mut a = [0.0; 3];
        for i in 0..3 {
            u[i] = bu[i] + g * du[i];
            a[i] = ba[i] + g * da[i];
        }
        (u, a)
    }

    /// Velocity of the composite at time 0.
    pub fn sample(&self, x: f64, y: f64, z: f64) -> [f64; 3] {
        self.sample_uw(x, y, z, 0.0).0
    }

    /// Vorticity of the composite.
    pub fn vorticity(&self, x: f64, y: f64, z: f64, t: f64) -> [f64; 3] {
        self.sample_uw(x, y, z, t).1
    }

    /// Helicity density `u·w` of the composite.
    pub fn helicity_density(&self, x: f64, y: f64, z: f64, t: f64) -> f64 {
        let (u, w) = self.sample_uw(x, y, z, t);
        u[0] * w[0] + u[1] * w[1] + u[2] * w[2]
    }

    /// The composite's analytic vector potential.
    pub fn potential(&self, x: f64, y: f64, z: f64, t: f64) -> [f64; 3] {
        self.sample_ua(x, y, z, t).1
    }
}

impl VectorPotential for TwoScale<'_> {
    fn velocity_and_potential(&self, x: f64, y: f64, z: f64, t: f64) -> ([f64; 3], [f64; 3]) {
        self.sample_ua(x, y, z, t)
    }
}


/// Options for [`exact_ns`].
pub struct ExactNsOptions {
    /// Shell wavenumber `k0` (used for both `kmin` and `kmax`).
    pub k0: f64,
    /// Kinematic viscosity `nu` (becomes `decay`); `0` gives the steady Euler member.
    pub nu: f64,
    /// Chirality `+1` or `-1` (becomes `helicity`, so every mode takes that sign).
    pub sign: f64,
}

impl Default for ExactNsOptions {
    fn default() -> Self {
        ExactNsOptions { k0: 2.0, nu: 0.02, sign: 1.0 }
    }
}

/// Options for a field that is an **exact solution of the Navier–Stokes equations**.
///
/// All modes share one `|k| = k0` and one handedness, which makes the field Beltrami
/// (`curl u = ±k0·u`); the nonlinear term is then a pure gradient absorbed by the pressure, so
/// the flow simply decays as `e^(-nu·k0²·t)` — exactly what `decay` does to every amplitude.
///
/// ```
/// use helix_noise::{exact_ns, ExactNsOptions, HelixField, HelixOptions};
/// let f = HelixField::create(HelixOptions { seed: 7, ..exact_ns(ExactNsOptions::default()) });
/// ```
pub fn exact_ns(o: ExactNsOptions) -> HelixOptions {
    HelixOptions {
        kmin: o.k0,
        kmax: o.k0,        // one shell: |k| is the same for every mode
        helicity: o.sign,  // p = ±1 makes every drawn sign that sign
        churn: 0.0,        // om = 0; viscous decay is the only time dependence
        decay: o.nu,
        tileable: false,
        ellipticity: 1.0, // Beltrami — the whole point
        ..Default::default()
    }
}

/// Energy-weighted polarization of a measured turbulence state.
pub struct NsTargets {
    /// Linear-polarization degree.
    pub d: f64,
    /// Per-shell helical fraction `|p|`.
    pub abs_p: f64,
    /// Its signed mean.
    pub signed_p: f64,
}

/// Measured developed-turbulence polarization.
pub const NS_TARGETS_DEV: NsTargets = NsTargets { d: 1.155, abs_p: 0.462, signed_p: 0.057 };
/// Measured forced-turbulence polarization.
pub const NS_TARGETS_FORCED: NsTargets = NsTargets { d: 1.062, abs_p: 0.516, signed_p: 0.229 };

/// Invert the per-mode relative helicity `2e/(1+e²) = p` for `e`.
fn eps_for_p(p: f64) -> f64 {
    (1.0 - (1.0 - p * p).sqrt()) / p
}

fn ns_bundle(t: &NsTargets) -> HelixOptions {
    HelixOptions {
        modes: 96,
        layout: crate::constants::Layout::Random, // a genuine ensemble, as the target was measured
        kmin: 1.0,
        kmax: 15.0, // the band holding ~99% of the measured energy
        ellipticity: eps_for_p(t.abs_p),
        helicity: t.signed_p / t.abs_p,
        ..Default::default()
    }
}

/// Option bundle whose **polarization** matches measured developed turbulence: per-mode helical
/// fraction `|p| ≈ 0.46` with a weakly positive signed mean. Both numbers are inverted exactly.
///
/// Scope, honestly: the polarization is calibrated, the **spectrum is not** — this keeps the
/// default power law. Supply your own `spectrum` if you have a measured one.
pub fn ns_developed() -> HelixOptions {
    ns_bundle(&NS_TARGETS_DEV)
}

/// As [`ns_developed`] for the *forced* state: more strongly polarized waves (`|p| ≈ 0.52`) and a
/// markedly net-handed field, which reads as a visible overall swirl direction. Same scope caveat.
pub fn ns_forced() -> HelixOptions {
    ns_bundle(&NS_TARGETS_FORCED)
}
